Transformations: Improve timeseries support

ryantxu opened the pull request:
When using the CalculateFieldTransformer with timeseries data, the results can be confusing unless you first join by the time field.

This PR aims to smooth the experience of using this with time values:
1. if the results are all timeseries, it will join them automatically first
2. if timeseries, it will keep the time field even when you specify "replaceAllFields"
3. Adds two new modes to help the most standard math requests:
* binary operations between two fields
* operation between a field and a numeric constant


![image](https://user-images.githubusercontent.com/705951/80455567-da047b00-88e0-11ea-8480-e40c3182e0f5.png)

![image](https://user-images.githubusercontent.com/705951/80455672-0f10cd80-88e1-11ea-99c0-d0512dd66310.png)

![image](https://user-images.githubusercontent.com/705951/80455785-3e273f00-88e1-11ea-8414-71204d850dfd.png)

This still needs better tests and some UI fixes...

ryantxu committed:
TestData/Graph: load arrow and zoom to data range (#23764)

ryantxu committed:
extract out the field creation parts

ryantxu committed:
extract out the field creation parts

ryantxu committed:
three math modes

ryantxu committed:
better timeseries support

ryantxu commented:
@daniellee -- thoughts on names, structure?

torkelo commented on packages/grafana-data/src/transformations/transformers/calculateField.ts:
```diff
@@ -1,82 +1,261 @@
+/**
+ * Return true if every frame has a time column
+ */
+function findTimeSeriesName(data: DataFrame[]): string | undefined {
```
The description is a bit off, returns true if frame has time column, but return type is string | undefined. 

The name is a bit confusing as well. getConsistentTimeFieldName ?

ryantxu replied:
ha!  guess what the first version of the function did?  but since the `join` transformer we have works on the name, this had to find the name field and can not just assume "time"

torkelo replied:
yes, but it does not return the name of the time series but the time field.

aknuds1 committed:
Docs: Fix building of docs (#23923)

* Docs: Fix building of docs
* CircleCI: Fixate grafana/docs-base image revision in job for building docs

mckn committed:
Docs: enable packages reference docs for 7-beta (#23953)

* added packages reference menu item.

* removed the draft flag.

* Updated docs by running script.

torkelo commented on packages/grafana-ui/src/components/TransformersUI/CalculateFieldTransformerEditor.tsx:
```diff
@@ -24,14 +50,21 @@ interface CalculateFieldTransformerEditorState {
 }
 
+const calculationModes = [
+  { value: CalculateFieldMode.ReduceRow, label: 'Reduce Row' },
+  { value: CalculateFieldMode.BinaryOperation, label: 'Binary Operation' },
+  { value: CalculateFieldMode.Scale, label: 'Scale Field' },
```
Please rename to Reduce row (lowercase r) and the rest. We are trying to align all buttons and headings to use sentence case.

torkelo reviewed (commented):
This is great stuff!

mckn committed:
AlertTab: some ui updates (#23971)

* updated the alerting tab.

* changed so we use a confirm button.

* removed uncommeneted import.

* Change to secondary buttons

Co-Authored-By: Dominik Prokop <dominik.prokop@grafana.com>

* trying to fix issue with panel of undefined.

* Fix prettier

* Update public/app/features/alerting/AlertTab.tsx

Co-authored-by: Dominik Prokop <dominik.prokop@grafana.com>

ivanahuckova committed:
Docs: Query history 7.0 updates  (#23955)

* Update docs about query history

* Update docs/sources/features/explore/index.md

Co-Authored-By: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

Dieterbe committed:
Metrictank: Fix meta inspector consolidator field names (#23838)

to match https://github.com/grafana/metrictank/pull/1798

aknuds1 committed:
Chore: Update Grafana version (#23985)

* Update Grafana version

torkelo committed:
Docs: What's new in 7.0 placeholder (#23987)

* Docs: What's new in 7.0 placeholder

* Updated makefile

Clarity-89 committed:
Search: minor fixes (#23984)

* Search: Use folder id as key when present

* Search: Do not render modals if not open

sakjur committed:
Enterprise: List 7.0 features (#23956)

aknuds1 committed:
CircleCI: Fix triggering of jobs for releases (#23999)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

marefr committed:
Fix pagination of issues/PR's in changelog generator (#23997)

Fix pagination of issues/PR's in changelog generator

Clarity-89 committed:
Search: Convert time pickers to CSF (#24002)

Commit:
updated docs for reporting (#23733)

* updated docs

* peering comments

* Added info about what version test mails requires

aocenas committed:
Tracing: Fix view bounds after trace change (#23994)

torkelo committed:
Docs: fix image link (#24011)

torkelo committed:
Update whats new (#24012)

aknuds1 committed:
Chore: Put what's new and release notes URLs in package.json (#24006)

* Put what's new and release notes URLs in package.json
* Upgrade build pipeline tool

marefr committed:
Update changelog for v7.0.0-beta1 (#24007)

Co-Authored-By: Marcus Efraimsson <marcus.efraimsson@gmail.com>
Co-Authored-By: Andrej Ocenas <mr.ocenas@gmail.com>
Co-Authored-By: Hugo Häggmark <hugo.haggmark@grafana.com>
Co-authored-by: Tobias Skarhed <1438972+tskarhed@users.noreply.github.com>

aknuds1 committed:
verify-repo-update: Fix Dockerfile.deb (#24030)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
CircleCI: Upgrade build pipeline tool (#24021)

* CircleCI: Upgrade build pipeline tool

xlson committed:
Devenv: ignore enterprise (#24037)

ivanahuckova committed:
Add header icon to Add data source page (#24033)

torkelo commented:
Sorry would love to help refine the UI but short pressed on time with release, bugs, docs, changelog, marketing etc all needing stuff 

Is there any breaking changes in this (for users who start using the current version of the calculate field transform in beta1)? Maybe we can wait a bit with this. and @dprokop could help refine the UI (if needed) later next week.

aknuds1 committed:
latest.json: Update testing version (#24038)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

alexanderzobnin committed:
Fix login page redirected from password reset (#24032)

tskarhed committed:
Storybook: Rewrite stories to CSF (#23989)

* ColorPicker to CSF format

* Convert stories to CSF

* Do not export ClipboardButton

* Update ConfirmButton

* Remove unused imports

* Fix feedback

xlson committed:
changelog enterprise 7.0.0-beta1 (#24039)

aknuds1 committed:
CircleCI: Bump grafana/build-container revision (#24043)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

dprokop commented:
happy to help

torkelo committed:
Changelog: Updates changelog with more feature details (#24040)

* Changelog: Updates changelog with more feature details

* spell fix

* spell fix

* Updates

* Readme update

* Updates

mckn committed:
Select: fixes so component loses focus on selecting value or pressing outside of input. (#24008)

* changed the value container to a class component to get it to work with focus (maybe something with context?).

* added e2e tests to verify that the select focus is working as it should.

* fixed according to feedback.

* updated snapshot.

xlson committed:
Devenv: add remote renderer to grafana (#24050)

dprokop committed:
NewPanelEditor: minor UI twekas (#24042)

* Forward ref for tabs, use html props

* Inspect:  add inspect label to drawer title

* Add tooltips to sidebar pane tabs, copy changes

* Remove unused import

* Place tooltips over tabs

* Inspector: dont show transformations select if there is only one data frame

* Review

aknuds1 committed:
Changelog: Add a breaking change (#24051)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
CircleCI: Unpin grafana/docs-base (#24054)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

Clarity-89 committed:
Search: close overlay on Esc press (#24003)

* Search: Close on Esc

* Search: Increase bottom padding for the last item in section

* Search: Move closing search to keybindingsSrv

* Search: Fix folder view

* Search: Do not move folders if already in folder

torkelo committed:
Docs: Adds deprecation notice to changelog and docs for scripted dashboards (#24060)

daniellee commented:
@ryantxu this was much easier to use with the Graph panel and the math part is such a killer feature. Have to test more and think more about the UI part of it. (Still seeing a bug with alias on the graph panel - it works on the table panel)

RichiH committed:
Update CHANGELOG.md (#24047)

Fix typo

Co-authored-by: Daniel Lee <dan.limerick@gmail.com>

joe-elliott committed:
Documentation: Alternative Team Sync Wording (#23960)

* Alternative wording for team sync docs

Signed-off-by: Joe Elliott <number101010@gmail.com>

* Update docs/sources/auth/team-sync.md

Co-Authored-By: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

mjtrangoni committed:
Fix misspell issues (#23905)

* Fix misspell issues

See,
$ golangci-lint run --timeout 10m --disable-all -E misspell ./...

Signed-off-by: Mario Trangoni <mjtrangoni@gmail.com>

* Fix codespell issues

See,
$ codespell -S './.git*' -L 'uint,thru,pres,unknwon,serie,referer,uptodate,durationm'

Signed-off-by: Mario Trangoni <mjtrangoni@gmail.com>

* ci please?

* non-empty commit - ci?

* Trigger build

Co-authored-by: bergquist <carl.bergquist@gmail.com>
Co-authored-by: Kyle Brandt <kyle@grafana.com>

ryantxu committed:
Merge remote-tracking branch 'origin/master' into binary-math

ryantxu committed:
more tests

ryantxu committed:
remove FieldConfig setting

ryantxu commented:
@dprokop -- can you take a look?  This is significantly better than what is in the alpha (does not explode for time series!) and supports the most common operation field1/field2 = %

I'm not sure what would make the UI much better... but any ideas would be great!  maybe getting the operator to style differently than the names?

dprokop commented:
@ryantxu I am gonna have a look at this ltr today. Will be awesome if we included this in stable!

ryantxu committed:
merged binary and reduce

bergquist commented:
> Disclaimer. beer(s) was consumed before writing this.

Writing down my first impressions of using transformation at all. In this case, using the graph panel.
I'm using `devenv` promtheus with  #A `go_goroutines` and #B `go_goroutines * 2` queries.

* Love the debugger!
* Fields are mentions a lot. Timeseries don't have fields. They have dimensions/labels/tags. 

### Reduce
* `Reduce` mentions rows. Whats rows in this case. none of the major TSDB's have rows.
* Selecting that reducer removes everything. 

### Filter by name
I googled regexps, as one does when entering regexps, but neither `sourbeer` or `/sourbeer/g` removed any time series. What does Time (3) mean?

### Filter by refId
I guess this could be useful in some scenarios with multiple chained transformations. But I don't understand when. Grafana already supports hiding queries. 

### Organize fields
Having the possibility to renames all the series returned from the query sounds nice. But it only adds the value to the end of the serie name.

### Join by field
I don't get this one :(

![image](https://user-images.githubusercontent.com/618863/80753019-dcf3ab80-8b2c-11ea-9db1-51beb4fc82ec.png)


### Add field from calculation
This is the juice! :+1: 
I would expect `Count` to count data points and not series but.
When I pick `Difference` all points are below zero. I probably don't understand what this function does... What first and last value?

### Labels to fields

![image](https://user-images.githubusercontent.com/618863/80753063-ebda5e00-8b2c-11ea-99ec-08b48782d1ba.png)

torkelo commented:
Thanks for the testing , transforms in this current early iteration is mainly for the table and needs more work to handle time
Series Better (we have already PRs open to improve it)

But it’s never going to be perfect, I can a million sql queries that would result in confusing messy graphs 

Why would you expect all data transforms to make perfect sense in a graph? 

Like many many graphite functions they only make sense in specific scenarios and for specific inputs. So we need some more detailed descriptions that is clear. But we just got it working , this  is v0.1 :) , let it be a little ruff around the edges! 

We need a quick way to toggle graph to table more easily as well to understand what the result is.

But the biggest problem is the time series and labels handling in table/data frame world, super difficult to get this correct and behaving in intuitive and good ways (especially with backward compatibility) 

Ryan and myself have Been working on much better labels support in Grafana (for the first time!!) labels will not just an internal in datasources use to evaluate series names in alias formats but actually used by Grafana  directly https://github.com/grafana/grafana/pull/24024  but it’s really hard to make it work for all cases and all scenarios

bergquist commented:
> transforms in this current early iteration is mainly for the table

I get that this is an early iteration and it _sounds great for the table panel_. According to usage stats ~77% of all data sources are time-series databases. We don't have any stats for panel usage (we should be able to build something for that like I did for alerting stats) but I think its fair to say that the graph panel is the bread and butter of Grafana.

Right now 2 out of 7 transformations work on time series (I only think field calculation is worth talking about) in the graph panel. 

So how do we communicate that this feature is focused on the table panel and mostly for row-based data sources? Cause right now I think most people who use Grafana for monitoring would find this feature quite confusing and useless, **as it is now**. 

> But we just got it working , this is v0.1 :) , let it be a little ruff around the edges!

Transforms have been a requested feature for a long time and we now place it next to the query tab which gives it a lot of attention. I think we are underdelivering for the common use-case by a landslide. The _very least_ we can do is to place info texts about the state of transforms in the graph panel somehow or disable it with a feature flag for graph panels. This feature has a lot of potentials and I think it's sad that this will be people's first impression. (using it in graph panel with the popular data sources)

Sorry if I sound harsh. I'm writing this from the kitchen as I'm juggling the kids :) Please assume positive intent.

torkelo commented:
Think the description texts for the transforms needs to be much expanded and to add more detail and wording that is relevant for both TSDBs and other data sources. 

And you are completely right this feature should have a beta notice at the top, I  think that is clear :) asked for help with that from UX team, struggling finding a design for a help box that looks good.

But there is only 24 hours in a day and working 16 of them every day for last 7 weeks without a day off has still left a few things for beta period. Sorry for minor frustrations, It’s just I know it’s not perfect and can be improved and is lacking docs etc :) 

But the feedback is useful and shows we have a bit more work todo than we thought. But also appreciate that what we are trying to do is not easy and will require a few mistakes and imperfections before we get it right.

kylebrandt commented:
A concept in GEL is that joins happen on the math operators. Currently this is based on Labels with some compatibility rules, so will need expansion for named table like responses. In the context of a math formula, my idea is that you can have different join options as an option/context to the formula.

So for example `$A * $B` can have an option for how A and B are joined. These sort of options in the context could also things like NaN handling and other stuff.

The goal is that you can "do the right thing" for common cases, but have options (and/or advanced options) to change behavior in the context of that math node.

marefr commented:
Not sure this is related to these changes or bug in reduce transformer?

I have a table with 2 columns:
![image](https://user-images.githubusercontent.com/1668778/80819089-a1142100-8bd4-11ea-8131-a4f0be688c7b.png)

Apply reduce transformation:
![image](https://user-images.githubusercontent.com/1668778/80819115-a96c5c00-8bd4-11ea-8ea4-5093d329048a.png)

Would have expected 1 row as result instead of 2. Maybe would have been fine if there was a way to remove that empty row. Resulting row should have Field with a value of `@hostname Sum` or similar.

ryantxu commented:
I'm struggling to follow what is suggested for this PR vs transformers in general.  We can all agree that transformers needs improvement... the goal with this PR is to make simple math work better for the most common case.

ryantxu committed:
Merge remote-tracking branch 'origin/master' into binary-math

ryantxu committed:
improve tests

bergquist commented:
> I'm struggling to follow what is suggested for this PR vs transformers in general.

Sorry about that. I guess this PR felt like the best place to dump my opinions.

> We can all agree that transformers needs improvement... the goal with this PR is to make simple math work better for the most common case.

I think simple math is good enough to be shipped but we should be clear about this feature's capability and think about how we communicate it. https://github.com/grafana/grafana/pull/24158 is a great step in that direction.

marefr commented:
> I'm struggling to follow what is suggested for this PR vs transformers in general.

Sorry

One improvement I would like to see is the ability to select "all fields". In the following example if I would like to do math of all fields, but each field is dynamic like server hosts, I cannot find a way to do that except selecting each field: 
![image](https://user-images.githubusercontent.com/1668778/80911690-41cf2180-8d38-11ea-8022-097a0a432abe.png)

Maybe `All fields` option or allow regular expressions?

ryantxu committed:
Merge remote-tracking branch 'origin/master' into binary-math

ryantxu commented:
@marefr -- the other calculation mode "Reduce row" will take everything on the same row and send it to the existing set of reducers.  The default is to send everything, or you can limit it to some of the number fields:

![image](https://user-images.githubusercontent.com/705951/80938552-6655e880-8d8e-11ea-9201-c847c95c26e6.png)

ryantxu committed:
update options after values change

ryantxu commented:
@torkelo -- I think this is ready to go... the styling could use some love, but this is a *significant* improvement over what is in beta1, and the more feedback we get with it the better :)

torkelo committed:
Minor refactoring and polish to UI

torkelo committed:
Minor fixes
